core/src/main/scala/algebra/ring/Ring.scala

```diff
@@ -32,11 +32,11 @@ trait Ring[@sp(Int, Long, Float, Double) A] extends Any with Rig[A] with Rng[A]
   def fromInt(n: Int): A = sumN(one, n)
 }
 
-trait RingFunctions {
+trait RingFunctions extends AdditiveGroupFunctions with MultiplicativeMonoidFunctions {
```

Comment:
I think if we had `CommutativeRing[A]` and `Ring[A]` at the same implicit precedence, it would create an ambiguity (and not compile). For this reason, I think that we usually work hard to ensure this isn't possible.

That said, parameterizing the trait on the type class type is not a bad idea -- but we would want to make sure we're doing it everywhere that is appropriate, not just here.